metsim/metsim.py

```diff
@@ -154,94 +158,220 @@ def load(self, job_list: list):
-            logger.info("Processing {}".format(locs))
-            ds = self.met_data.isel(**locs)
-            lat = ds['lat'].values
-            elev = ds['elev'].values
-            df = ds.drop(['lat', 'lon', 'elev']).to_dataframe()
-            df = self.method.run(df, MetSim.params, elev=elev,
-                                 lat=lat, disagg=self.disagg)
-            self._unpack_results((locs, df))
+        time_dim = pd.to_datetime(self.met_data.time.values)
+        if self.params['annual']:
+            groups = time_dim.groupby(time_dim.year)
+        else:
+            groups = {'total': time_dim}
+        for label, times in groups.items():
+            logger.info("Beginning {}".format(label))
+            # Add in some end point data for continuity
+            times_ext = times.union([times[0] - pd.Timedelta("1 days"),
+                                     times[-1] + pd.Timedelta("1 days")]
+                                    ).intersection(time_dim)
+            data = self.met_data.sel(time=times_ext)
+            self.setup_output(self.met_data.sel(time=times))
+            for i, j in self.locations:
+                locs = dict(lat=i, lon=j)
+                logger.info("Processing {}".format(locs))
+                ds = data.isel(**locs)
+                lat = ds['lat'].values
+                elev = ds['elev'].values
+                swe = ds['swe'].values
+                df = ds.drop(['lat', 'lon', 'elev', 'swe']).to_dataframe()
+                # solar_geom returns a tuple due to restrictions of numba
+                # for clarity we convert it to a dictionary here
+                sg = solar_geom(elev, lat)
+                sg = {'tiny_rad_fract': sg[0], 'daylength': sg[1],
+                      'potrad': sg[2], 'tt_max0': sg[3]}
+
+                df = self.method.run(df, self.params, sg,
+                                     elev=elev, swe=swe)
+
+                if self.disagg:
+                    try:
+                        prevday = data.time[0] - pd.Timedelta('1 days')
+                        t_begin = [self.met_data['t_min'].sel(
+                                       time=prevday).isel(lat=i, lon=j),
+                                   self.met_data['t_max'].sel(
+                                       time=prevday).isel(lat=i, lon=j)]
+                    except (KeyError, ValueError):
+                        t_begin = [self.state['t_min'].values[-1, i, j],
+                                   self.state['t_max'].values[-1, i, j]]
+                    try:
+                        nextday = pd.datetime(int(label)+1, 1, 1)
+                        t_end = [self.met_data['t_min'].sel(
+                                     time=nextday).isel(lat=i, lon=j),
+                                 self.met_data['t_max'].sel(
+                                     time=nextday).isel(lat=i, lon=j)]
+                    except (KeyError, ValueError):
+                        t_end = None
```

Comment:
Can you add some comments to this section. I'm a bit confused why `t_begin` and `t_end` need to be lists `[tmin, tmax]` and the try except statements are a bit tough to follow so some comments would be useful.